File: memory.py
#!/usr/bin/env python3

import sys
import os.path
from os import path
import note
import helper_func as hf

# Color Definitions
BOLD = '\033[1m'
ITALIC = '\033[3m'
RESET = '\033[0m'


"""
Given a memory and the name of a 
note, appends the memory to the
end of the note
"""
def append_memory(memory, note):
	with open(note, 'a') as f:
		memory += '\n'
		f.write(memory)


"""
Given the current_note and a row number,
removes the memory at that row and returns
its value
"""
def change_memory(current_note, row):
	memories = hf.check_row(current_note, row)	
	if memories == None:
		return
	row = int(row)
	memory_to_remove = memories[row - 1].strip('\n')
	del memories[row - 1]
	note.write_to_note(current_note, memories)
	return memory_to_remove


"""
Given the name of a note and
a row number, moves the memory
at that row number in the current
note to the given note
"""
def move_memory(current_note, row, args):
	note = hf.parse_unary_args(args)
	note_path = path.expanduser('~/.quicknote/.notes/.' + note)
	if not path.isfile(note_path):
		print('The note you are trying to add to does not exist. Please try again.')	
		return
	memory_to_move = change_memory(current_note, row)	
	if memory_to_move == None:
		return
	append_memory(memory_to_move, note_path)
	print('Moved \'' + memory_to_move + '\' to \'' + note + '\'')
		

"""
Given the name of a note and
a row number, copies the memory
at that row number in the current
note to the given note
"""
def copy_memory(current_note, row, args):
	note = hf.parse_unary_args(args)
	note_path = path.expanduser('~/.quicknote/.notes/.' + note)
	if not path.isfile(note_path):
		print('The note you are trying to add to does not exist. Please try again.')
		return
	memories = hf.check_row(current_note, row)	
	if memories == None:
		return
	row = int(row)
	memory_to_copy = memories[row - 1].strip('\n')
	append_memory(memory_to_copy, note_path)
	print('Copied \'' + memory_to_copy + '\' to \'' + note + '\'')


"""
Given a note and a row number,
removes the memory at that row number
"""
def remove_memory(note, row):
	if not path.isfile(note):
		return
	memory_to_remove = change_memory(note, row)	
	if memory_to_remove == None:
		return
	print('Removed memory \'' + memory_to_remove + '\'')


"""
Given a filename, removes all
memories from that file
"""
def clear_memories(filename):
	prompt = 'Are you sure you want to clear your current memories on this note?' + BOLD + ' There is no going back (y/n): ' + RESET
	decision = hf.request_user_permission(prompt)
	if decision == 'y':
		open(filename, 'w').close()
		print('Memories cleared')
	elif decision == 'n': # this probably isn't necessary (i.e. don't need to check, could be an else)
		print('Memory clearing aborted')	


"""
Given a filename and a list of words,
adds those words (as a single line) to
the end of the file
"""
def add_memory(note, args):
	memory = hf.parse_unary_args(args)
	append_memory(memory, note)
	print('Remembered \'' + memory + '\'')


"""
Given a filename, lists the current memories
held by that file
"""
def list_memories(filename):
	if not path.isfile(filename):
		print('Hmmm. The current note doesn\'t seem to be working. Please try again later.')
		print('If the current note continues to fail to open, please submit a help ticket by emailing us at quicknote.v1@gmail.com')
		return
	elif path.getsize(filename) == 0:
		print('You have no memories on this note')
		return
	with open(filename, 'r') as f:
		counter = 1
		for memory in f:
			memory = str(counter) + '. ' + memory 
			print(memory, end='')
			counter += 1


"""
Given a row number, removes the
memory at that row and adds it to
the '.archive' memory file
"""
def archive_memory(filename, row):
	archive_note = path.expanduser('~/.quicknote/.notes/.archive')
	memories = hf.check_row(filename, row)
	if memories == None:
		return
	row = int(row)
	memory_to_archive = memories[row - 1]
	del memories[row - 1]
	with open(filename, 'w') as f:
		for memory in memories:
			f.write(memory)
	with open(archive_note, 'a') as f:
		f.write(memory_to_archive)
	print('Archived memory \'' + memory_to_archive.strip('\n') + '\'')

File: C_implementation/memory.c
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>
#include "memory.h"
#include "helper_func.h"
#include "note.h"
// #include "../../json-c/json_object.h"

void appendMemory(char* memory, char* note) {
	FILE *toWrite;
	toWrite = fopen(note, "a");
	fprintf(toWrite, "%s\n", memory);
	fclose(toWrite);
}


int changeMemory(char* charRow, char* memory) {
	char currentNoteName[256];
	char currentNotePath[256];
	getCurrentNote(currentNotePath, currentNoteName);
	int result = checkRow(currentNotePath, charRow);
	if (result == -1) return -1;
	
	int row = atoi(charRow);
	
	FILE* source;
	source = fopen(currentNotePath, "r");

	FILE* target;
	char dirName[] = "background";
	char noteName[] = "BRIEF";
	char notePath[256];
	getNotePath(dirName, noteName, notePath);
	target = fopen(notePath, "w");

	char temp[256];
	char line[256];
	int counter = 0;

	while (fgets(line, sizeof(line), source)) {
		if (counter + 1 == row) {
			strcpy(temp, line);
			counter++;
			continue;
		}
		fprintf(target, "%s", line);
		counter++;
	}
	
	// the following is slightly odd but seems to work	
	int sizeMemory = strlen(temp);
	strncpy(memory, temp, sizeMemory - 1);
	memory[sizeMemory - 1] = '\0';	

	fclose(source);
	fclose(target);

	remove(currentNotePath);
	rename(notePath, currentNotePath);

	//writeToNote(currentNotePath, lines, counter);
	return 0;	
	
}


void moveMemory(char* args[], int numArgs) {
	char note[256];
	parseSpecialArgs(note, args, numArgs);
	
	char dirName[] = "notes";
	char notePath[256];
	getNotePath(dirName, note, notePath);
	
	struct stat st = {0};
	if (stat(notePath, &st) == -1) {
		printf("The note you are trying to add to does not exist. Please try again.\n");
		return;
	}
	int result = checkDefault(note);
	if (result == -1) return;

	char memoryToMove[256];
	result = changeMemory(args[2], memoryToMove);
	if (result == -1) return;

	appendMemory(memoryToMove, notePath);
	printf("Moved '%s' to '%s'\n", memoryToMove, note);
}


void copyMemory(char* args[], int numArgs) {
	char note[256];
	parseSpecialArgs(note, args, numArgs);

	char currentNoteName[256];
	char currentNotePath[256];
	getCurrentNote(currentNotePath, currentNoteName);

	char dirName[] = "notes";
	char notePath[256];
	getNotePath(dirName, note, notePath);
	
	struct stat st = {0};
	if (stat(notePath, &st) == -1) {
		printf("The note you are trying to add to does not exist. Please try again.\n");
		return;
	}
	int result = checkDefault(note);
	if (result == -1) return;

	
	result = checkRow(currentNotePath, args[2]);
	if (result == -1) return; 
	
	int row = atoi(args[2]);

	FILE* source = fopen(currentNotePath, "r");
	int counter = 0;
	
	char line[256];
	char temp[256];
	while (fgets(line, sizeof(line), source)) {
		if (counter + 1 == row) {
			strcpy(temp, line);
			counter++;
			continue;
		}
		counter++;
	}

	fclose(source);

	char memoryToCopy[256];

	// the following is slightly odd but seems to work	
	int sizeMemory = strlen(temp);
	strncpy(memoryToCopy, temp, sizeMemory - 1);
	memoryToCopy[sizeMemory - 1] = '\0';	

	appendMemory(memoryToCopy, notePath);
	printf("Copied '%s' to '%s'\n", memoryToCopy, note);
}


void removeMemory(char* row) {
	char memoryToRemove[256];
	int result = changeMemory(row, memoryToRemove);
	if (result == -1) return;

	printf("Removed memory '%s'\n", memoryToRemove);
}


void clearMemories() {
	char currentNotePath[256];
	char note[256];
	getCurrentNote(currentNotePath, note);

	char prompt[] = "Are you sure you want to clear your memories on your current note? \033[1mThere is no going back (y/n): \033[0m";
	char decision[50];
	requestUserPermission(prompt, decision);
	if (strcmp(decision, "y") == 0) {
		fclose(fopen(currentNotePath, "w"));
		printf("Memories cleared\n");
	} else printf("Memory clearing aborted\n");	
}


void addMemory(char* args[], int numArgs) {
	char note[256];
	char currentNotePath[256];
	getCurrentNote(note, currentNotePath);
	char sansFirstWord[256];
	int result = parseUnaryArgs(sansFirstWord, args, numArgs);
	if (result == 1) {
		appendMemory(args[1], note);
		printf("Remembered '%s'\n", args[1]);
	} else {
		char memory[256];
		// The following is because the program considers the first word after program call to be the 'command'
		// This word must be added back
		strcpy(memory, args[1]);
		strcat(memory, " ");
		strcat(memory, sansFirstWord);

		appendMemory(memory, note);
		printf("Remembered '%s'\n", memory);
	}
}


void listMemories() {
	char notePath[256];
	char noteName[256];
	getCurrentNote(notePath, noteName);

	if (access(notePath, F_OK) == -1) {
		printf("Hmmm. The current note (%s) doesn't seem to exist. How did we get here?\n", noteName);
		return;
	}	

	// get the number of lines in the file 
	int numLines = getFileSize(notePath);
	if (numLines == 0) {
		printf("You have no memories on this note\n");
		return;
	}
	printf("\033[1m\033[3mFound %d memories on '%s'\033[0m\n", numLines, noteName);

	FILE* toRead;
	toRead = fopen(notePath, "r");
	int counter = 1;
	char memory[256];

	while (fgets(memory, sizeof(memory), toRead)) {
		printf("\033[1m%d.\033[0m %s", counter, memory); 
		counter++;
	}	

	fclose(toRead);
}
	
